h2o-core/src/main/java/water/H2O.java

```diff
@@ -1619,10 +1620,6 @@ public H2ONode getNodeByIpPort(String ipPort) {
   H2O( H2ONode[] h2os, int hash, int idx ) {
     _memary = h2os;             // Need to clone?
     java.util.Arrays.sort(_memary);       // ... sorted!
-    _node_ip_to_index = new HashMap<>();
```

Comment:
Can you please explain why do you need to move the initialization?

Reply:
I moved it because when the Initialisation was here the map was empty for the client node.

My explanation is that the class needs to be fully initialised with all the nodes members before we can create the mapping. ( Because of the indexes calculation)

Reply:
_My explanation is that the class needs to be fully initialised with all the nodes members before we can create the mapping. ( Because of the indexes calculation)_

This is not true even after your change, or is it? New members can join and set_next_Cloud will be called again.

new H2O is called from 2 places: from set_next_Cloud and from

   public static volatile H2O CLOUD = new H2O(new H2ONode[0],0,0);

It seems to me that your change just moves the code but the code is still executed in the same order. What I am struggling here is to understand how can this fix anything.

Reply:
oki, in that case I don't see why this changes it as well. However it really fixes the situation when we try to get logs from client node. Before this fix the node mapping was empty. I will try to debug more to come up with proper explanation

Reply:
@michalkurka so thoughts behind the explanation above should be correct, i just explained it bad. Here's the explanation why this change actually works:
`new H2O(..)` constructor sets the new mapping from `node_ip_to_index` iteratively as 
` _node_ip_to_index.put(node.getIpPortString(), node.index());`, however the `node.index()` is calculated on the old cloud information `H2O.CLOUD` wasn't still updated and  `memary` doesn't contain the node. That's why we always get `-1` as the node port which fallbacks to the current node and the logs from the remote node are not showed up.

The changes moves the initialisation of `_node_ip_to_index` map to the code part where we updated H2O.CLOUD to a new version, therefore `memary` is updated and `node.index()` can return the proper index

Reply:
I see now. Thanks for explaining!